Rust generator for LinkML

kervel opened the pull request:
see https://github.com/linkml/linkml/issues/2360

this PR supersedes:
* https://github.com/linkml/linkml/pull/2364 
* https://github.com/linkml/linkml/pull/2722 

it adds an extra test workflow that pulls in the rust toolchain for the rust related tests.

kervel commented:
> ## [Codecov](https://app.codecov.io/gh/linkml/linkml/pull/2889?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=linkml) Report
> ❌ Patch coverage is `0%` with `1138 lines` in your changes missing coverage. Please review. ✅ Project coverage is 0.31%. Comparing base ([`9323b73`](https://app.codecov.io/gh/linkml/linkml/commit/9323b73ba548debe5015c10f445ca95c0c5066bb?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=linkml)) to head ([`84f37ed`](https://app.codecov.io/gh/linkml/linkml/commit/84f37ed46d34fb7889158f7ae860ac9c3b57ab83?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=linkml)).

i guess this has to do with my new tests being disabled still.. but i have no idea how to enable them for codecov.

sneakers-the-rat commented:
> SkipValidation rewrite looks broken. 

Do you mean the version in this PR is broken, or was the original broken? It was working fine afaik, why change it?

> GetPydanticSchema is not a public pydantic v2 API 

it is!
https://docs.pydantic.dev/latest/concepts/types/#using-getpydanticschema-to-reduce-boilerplate

> Using a TypeVar inside Annotated as a module-level alias (Tsv) and then subscripted later is not how Annotated/generics are intended to be used

It is!
https://typing.python.org/en/latest/spec/qualifiers.html#annotated

sneakers-the-rat commented on tests/test_generators/test_rustgen_maturin.py:
```diff
@@ -0,0 +1,128 @@
+def test_rustgen_personschema_maturin(input_path, temp_dir):
+    """
+    Generate a Rust crate from personinfo.yaml and try to build
+    the Python bindings with maturin. Skips if maturin/cargo are absent.
```
i think we shouldn't double-skip here: you already are skipping in the conftest, so if have reached this test we are explicitly intending the rustgen tests to run. so if we reach here and still don't have the appropriate dependencies, we should fail.

sneakers-the-rat commented on tests/test_generators/test_rustgen_maturin.py:
```diff
@@ -0,0 +1,128 @@
+    wheels_dir = out_dir / "target" / "wheels"
+    assert wheels_dir.exists(), "maturin did not produce target/wheels"
+    wheels = list(wheels_dir.glob("*.whl"))
+    assert wheels, "no wheel produced by maturin"
```
can we try and import the wheel here?

kervel replied:
ok, i'll try. its not as simple to import a binary wheel. but it can be done.

sneakers-the-rat commented on tests/test_generators/test_rustgen_maturin.py:
```diff
@@ -0,0 +1,128 @@
+    gen = RustGenerator(metamodel_sv.schema, mode="crate", pyo3=True, serde=True, output=str(out_dir))
+    gen.serialize(force=True)
+
+    cargo_toml = out_dir / "Cargo.toml"
+    pyproject_toml = out_dir / "pyproject.toml"
+    assert cargo_toml.exists(), "Cargo.toml not generated"
+    assert pyproject_toml.exists(), "pyproject.toml not generated"
```
sort of duplicating tests here. probably would be good to turn "build a rustgen schema" into a fixture so that we don't end up needing to copy/paste for every rustgen test.

sneakers-the-rat commented on examples/PersonSchema/rust/Cargo.toml:
in general i feel like any examples should get auto-built or else they're guaranteed to go out of date. at least a test that regenerates and confirms that the examples are correct. out of date examples are sometimes worse than no examples.

kervel replied:
you are right. There was a Makefile under personschema so i added the rust generation there. but it seems the Makefile is just sitting there, when i run "make" it seems also the pydantic and other examples were outdated. so probably this needs some extra thinking.

sneakers-the-rat reviewed (commented):
just commenting on common/testing things. dont have time to review actual rust gen atm. pr is pretty humongous

kervel commented:
> > SkipValidation rewrite looks broken.
> 
> Do you mean the version in this PR is broken, or was the original broken? It was working fine afaik, why change it?
> 
> > GetPydanticSchema is not a public pydantic v2 API
> 
> it is! https://docs.pydantic.dev/latest/concepts/types/#using-getpydanticschema-to-reduce-boilerplate
> 
> > Using a TypeVar inside Annotated as a module-level alias (Tsv) and then subscripted later is not how Annotated/generics are intended to be used
> 
> It is! https://typing.python.org/en/latest/spec/qualifiers.html#annotated

sorry about this! i used AI help to look at common.py because i was not deep into it enough myself to understand everything, and i tought i removed all AI rubbish before posting the comment, obviously i didn't...

kervel commented:
Hello @sneakers-the-rat

Thanks for the review! I think i addressed most of your issues, i don't know if the `common.py` now is good enough (see my comment). But if you think it is, i guess the PR is ready for another review (i'll ask who could be a candidate). Indeed reviewing properly will be tough because of the size of the PR..

sneakers-the-rat commented on linkml/generators/common/template.py:
```diff
@@ -42,7 +43,7 @@ def render(self, environment: Optional[Environment] = None, **kwargs) -> str:
         """
         if environment is None:
-            environment = TemplateModel.environment()
+            environment = type(self).environment()
```
I'll take a closer look, I only took a skim yesterday, but i can do a more complete check. I won't have time for a full review of the rustgen and would hope we could get a lot of the validation by plugging it into the core spec tests, though I don't know how we would make that not be ridiculously expensive to test since we would need to compile every test and there are hundreds/thousands.

tfliss commented:
@kervel Ok, I checked out and ran/passed the test. I've wanted to ramp up on rust, so am planning to to review in some detail. Is there a particular timeline / deadline you have in mind?

kervel commented:
> @kervel Ok, I checked out and ran/passed the test. I've wanted to ramp up on rust, so am planning to to review in some detail. Is there a particular timeline / deadline you have in mind?

hey,

not particulary.. my main concern is that if it stays too long in a separate branch it will become unmergeable/bitrotted.. 
also, i was planning to add wasm bindings too to the rust generated crate, but that would make the already big PR even bigger, so maybe i should do that in a separate PR.

if you want to check the rust side too, it might be handy to look at a crate that actually uses the generated classes. For this, https://github.com/Kapernikov/rust-linkml-core/tree/new-rustgen is a candidate ("new-rustgen" branch!). its a rust schemaview and some tools with python bindings.

thanks already!
frank

tfliss commented:
@kervel As far as the compliance tests that @sierra-moxon recommended, it might be worth working on that as a separate issue. I'll comment in my review how you may want to document the current state of implementation.

This [draft PR in a fork](https://github.com/linkml/linkml/pull/2895/files) contains my current work-in-progress PolaRS generator. The relevant part for you would be tests/test_compliance/ (please ignore the rest).

- in helper.py:
    - add a constant for your generator: `RUST_GEN = "rust_gen"`
    - add an entry with the class in the GENERATORS dict: `RUST_GEN: generators.RustGenerator,`
    - in check_data() delegate the check to your generator:
    ```
    if isinstance(gen, GENERATORS[RUST_GEN]):
            check_data_rustgen(.....)
    ```
    - add the definition of check_data_rustgen()
        - note that I'm using `pl = pytest.importorskip("polars", minversion="1.0", reason="Polars >= 1.0 not installed")` to conditionally import an optional package and skip the tests if it's not present.
        - I check `schema['name']` to skip individual tests rather than touching every test_compliance/test_*.ply file.
            - could also just skip anything except specifically enabled tests to start.
        - The body of this check is similar to your `test_rustgen_personschema_maturin()`
            - target_class is the name of your model
            - object_to_validate will be a dict
            - schema is the linkml yaml
            - expected_behavior indicates whether your generator implements the feature or not
            - valid is whether the object_to_validate is expected to pass or fail validation.
            - my simplistic approach is to try to round-trip the object_to_validate back to a dict and then compare them
            - note: I'm not necessarily recommending this approach, it's just my current work-in-progress and is probably something you could use to get a basic compliance test running.
- in test_compliance.py:
    - hook up the constant you defined in helper.py to a pytest mark in CORE_FRAMEWORKS: `pytest.param(RUST_GEN), marks=[pytest.mark.with_panderagen]),`
 - in test_core_compliance.py:
     - look at the `_mapping` parameters for examples of comparing your generated code to what is expected for the input schema (most of them are very small schemas)
     - I found stepping through with a debugger as helpful for some of these tests, they are heavily parametrized (allso see --with-output).
 
 At this point I generally invoke using something like the following (you should need to use your pytests mark):
 
 ```
uv run pytest -s -vv --log-cli-level=INFO tests/test_compliance -m with_rustgen --with-output -k test_attributes | tee test.out
```

The `--with-output` flag will drop detailed information about each test under tests/test_compliance/output/ in a directory structure reflecting the test_compliance modules, individual tests and the various generators. 
        
There is also a [test_compliance/README.md](https://github.com/linkml/linkml/blob/main/tests/test_compliance/README.md)  

I hope this helps get you started. Feel free to reach out if anything doesn't make sense or doesn't work. I wouldn't expect it to take much to get these tests hoooked up (probably not all passing, but some should). If you get stuck I may have missed a detail. I'm also very open to suggestions for other approaches as I am also learning my way around the compliance tests.

tfliss commented:
@kervel Here's a quick patch with mostly the same changes as described above, in this gist.: https://gist.github.com/tfliss/75462c9d5284973ce06f7f7773afc233

The run_one_test.sh script will run one parametrized enum test and generate output. 12 pass and 12 fail, which is great for the first try. I think some of the errors are due to needing name cleaning in a couple of places (schema name). The check is mostly a copy/paste from one of your tests, then instantiate the object and compare the top-level key/value pairs. Not production quality, but somewhere to start. Speed seems ok. Once the schema name thing is fixed we can run more of the suite.

Hope this helps, I can also push a branch or share the logs and generated code.

I'll continue the review next week. Have a great weekend. So far this has been great to work with. Thank you.

sneakers-the-rat commented:
I think this would be great to break up into smaller PRs. I would recommend no new features in this one, and focus on testing.

I suspect that the rust tests requiring compilation for each schema will make the tests take an enormous amount of time, though we should measure that, so they will likely need to be split out into their own job like the slow tests. 

The compliance tests are definitely the target we need to eventually hit, but it may require a significant amount of work, since there currently isn't a good way to do "partial compliance" without manually editing the skip conditions in the body of each test. In the meantime, the checklist in the scoping issue may be a useful guide for what would be good to have in rustgen-specific tests. Without those it's hard to properly review, since its hard to know what is implemented, what isn't, and whether what is written works.

The compliance tests are designed so there is one schema per test (mostly, except for some parameterizations), which would mean one compilation per test, but for many of the base-case rustgen boilerplate tests, we could probably get away with a module-scoped fixture that compiles the kitchen sink schema once and use that for everything.

tfliss reviewed (changes requested):
Overall to me this looks like a solid and useful initial generator implementation. Thank you @kervel 

I've looked over pretty much everything, with an assist from an LLM on the generated PersonSchema rust. Ran the makefile, ran a lot of compliance tests in a branch (some passed, some failed), coverage looks quite good. I did not go over the jinja2 templates in detail, but had eyes on it all and the style and readability are high quality.

If anything is split into a separate PR I'd like to see the linkml/generators/common changes be reviewed more widely, and that would be fairly small.

Other than that I'd want to see a note in the .rst with respect to general compliance testing being a work-in-progress. Also please look over the remaining TODO comments to check if they are still relevant, some commented out code fragments and the two functions that might be unused per coverage. If those changes are made I will approve this PR (but still recommend a final approval by a core dev).

Have a good weekend!

kervel commented on linkml/generators/rustgen/rustgen.py:
```diff
@@ -0,0 +1,1037 @@
+            else None
+        ),
+        child_ranges=sub_ranges if len(sub_ranges) > 1 else None,
+        box_needed=inline_mode == SlotInlineMode.INLINE and can_contain_reference_to_class(s, cls, sv),
```
hello, the boxing is needed to keep structs sized, and unfortunately it doesn't make the generated code more open-world ...

basically you can't have a Mystruct with a member of type Mystruct in rust, because that would lead to infinite size. The inner Mystruct needs to be allocated on the heap (hence the box)

see this: https://stackoverflow.com/questions/25296195/why-are-recursive-struct-types-illegal-in-rust

kervel commented on docs/generators/rust.rst:
```diff
@@ -0,0 +1,194 @@
+
+Overview
+--------
+
```
i have put something but i'm not sure if it is what you want.. my best "view" on the completeness of this is the list @sneakers-the-rat originally made in #2360

sneakers-the-rat commented:
it's already merged but fwiw changes to common/pydanticgen look fine. 

thanks for carrying this through @kervel :)
